### PR TITLE
add ListByIndex func to use index while list objects

### PR DESCRIPTION
**Pull request by adamzhoul**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

k8s have a very powerful design of indexer and generate a lot of  object listers, like `podLister`
And,  those `Lister` should be the most encouraging way to list the objects.
However,  those `Lister` can only make use of the namespace index to find objects.

so, we need a way to use an index to get a quick find.

#### The situation very useful  

In my situation:
- we have serval namespaces, each of them has hundreds of pods.
- our app always deployed multiple versions at the same time. 
- we should show people, all of the pod they have deployed with `labelSelector: "app=xxx"`

so, if we use default lister using namespace index, we have to find hundreds of pods one by one.
but, if we can self custom a new index, group by app name, and use it, it can be much quicker.

#### Special notes for your reviewer:

1. this pr is the first step, to achieve the easy part. add underlay function.
2. get great ideas from the community, see if change the list-generator template is the best way.
3. using informer.GetIndex can work too, but using lister is more encouraged I think. 

`Lister` code is generated and can't be modified by the user (PaaS/saas platform developer, who use `lister`) 
 https://github.com/kubernetes/code-generator/issues/84

#### Does this PR introduce a user-facing change?

```release-note
     using lister-generator will change the interface signature
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/listers.go [hunk 1]
```diff
@@ -108,6 +108,42 @@ func ListAllByNamespace(indexer Indexer, namespace string, selector labels.Selec
 	return nil
 }
 
+// ListByIndex used to list items using specific index
+func ListByIndex(indexer Indexer, name, key string, selector labels.Selector, appendFn AppendFunc) error {
+
```
Comment (lala123912) [inline staging/src/k8s.io/client-go/tools/cache/listers.go:113]: redundant blank line

File: staging/src/k8s.io/client-go/tools/cache/listers.go [hunk 2]
```diff
@@ -108,6 +108,42 @@ func ListAllByNamespace(indexer Indexer, namespace string, selector labels.Selec
...
+	items, err := indexer.ByIndex(name, key)
+	if err != nil {
+		// if err found, we should not list all , return err instead.
+		// if caller needs list all, do not need use specific index, use ListAll
+		klog.Warningf("can not retrieve list of objects using index : %v", err)
```
Comment (lala123912) [inline staging/src/k8s.io/client-go/tools/cache/listers.go:118]: use InfoS instead.

File: staging/src/k8s.io/client-go/tools/cache/listers_test.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
+package cache
```
Comment (lala123912) [inline staging/src/k8s.io/client-go/tools/cache/listers_test.go:1]: add copyright

File: staging/src/k8s.io/client-go/tools/cache/listers_test.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	fcache "k8s.io/client-go/tools/cache/testing"
+
+	"errors"
+	"fmt"
+	"testing"
+	"time"
```
Comment (lala123912) [inline staging/src/k8s.io/client-go/tools/cache/listers_test.go:13]: adjust the order of  import

File: staging/src/k8s.io/client-go/tools/cache/listers_test.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	WaitForCacheSync(stop, informer.HasSynced)
+
+	devEnvSelector, _ := labels.Parse("env=dev")
+	tests := []struct {
+		name      string
+		args      args
+		resources map[string]bool
+	}{
+		{
+			name: "hello-dev",
```
Comment (lala123912) [inline staging/src/k8s.io/client-go/tools/cache/listers_test.go:97]: the name of test case is not specified.

File: staging/src/k8s.io/client-go/tools/cache/listers.go (hunk 1 above)
Comment (adamzhoul) [reply staging/src/k8s.io/client-go/tools/cache/listers.go]: deleted

File: staging/src/k8s.io/client-go/tools/cache/listers.go (hunk 2 above)
Comment (adamzhoul) [reply staging/src/k8s.io/client-go/tools/cache/listers.go]: done

**Comment by caesarxuchao** [comment]
/assign @deads2k @yliaog 
Thank you!
/triage accepted

**Comment by yliaog** [comment]
have you checked out:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/listers/core/v1/pod.go#L33

it should be able to solve the use case described in the PR description

**Comment by adamzhoul** [comment]
thanks  @yliaog 

what do you mean checked out? 
```
// PodLister helps list Pods.
// All objects returned here must be treated as read-only.
type PodLister interface {
	// List lists all Pods in the indexer.
	// Objects returned here must be treated as read-only.
	List(selector labels.Selector) (ret []*v1.Pod, err error)
	// Pods returns an object that can list and get Pods.
	Pods(namespace string) PodNamespaceLister
	PodListerExpansion
}
```

- from the interface signature, we can only pass in a selector, can't specify which index to use.
- do you mean we implement this interface? instead of using the `podList`  the package provided?
- when we use informer to return a lister, it actually returns `podList`,  it just listAll or uses the namespace index.
- I don't know how to make use of `PodListerExpansion`, can it solve my problem. how? thanks.

we always list objects using the code k8s code-generator generated. 
isn't it better if the index function is natively supported?


thanks very much

**Comment by adamzhoul** [comment]
ping @yliaog,  any advice in specific?

thanks

**Comment by yliaog** [comment]
as far as i understand it, the use case in the PR description would need to find all the pods in all namespaces with labelSelector: "app=xxx"

if this is the case, then you don't need to specify which indexer to use, right?

**Comment by adamzhoul** [comment]
thanks for the reply.  @yliaog 

 1.  yes,  we can get the pods. But has to for-range all pods in all ns, and calculate label selector match.
     but, with the index, for-range all is not needed.  pass in label: app=helloworld, we get the pods directly.
     this is a kind of accelerate.


 2.  `postLister`  can only make use of the namespace index,  there should be a way to use a customer index. 
       not only limited in my situation.

  besides, features like `SelectorIndex`  [#92503](https://github.com/kubernetes/kubernetes/pull/92503) can use this func ListByIndex maybe.

**Comment by yliaog** [comment]
it's good to use a custom indexer to speed up, below is a quick example to show how it works.

It creates custom indexer that index pods by its node name.  then it is used to quickly find all the pods on a given node with: podInformer.Informer().GetIndexer().ByIndex("nodeName", node.Name)

// Create custom Indexer
podInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
		"nodeName": indexByPodNodeName,
	})

func indexByPodNodeName(obj interface{}) ([]string, error) {
	pod, ok := obj.(*v1.Pod)
	if !ok ||  len(pod.Spec.NodeName) == 0 {
		return []string{}, nil
	}
	return []string{pod.Spec.NodeName}, nil
}

**Comment by adamzhoul** [comment]
yes, your example clearly shows how we add and use an index now.  And I totally agree with that.

So, the pr here only makes it more convenient to use. 
And, I wish we can get objects by `Lister` not `informer`,  this is the intention why I add this.

finally thought could be like :

```golang
   
 // add index and start informer in controller
   podInformer := // 
   podInformer.AddIndex(...)


// use lister to get objects.
    lister := ... // lister object get by code-generator, we don't need to reimplement. it uses the informer 
    lister.ListByIndex(indexName,  labelSelector) // tells which index to use, and calcualte labelSelector if needed.

```


thanks @yliaog

**Comment by yliaog** [comment]
namespace is special in kubernetes, listing pods by namespace is very common, hence it is added to the client library. I think it's helpful to have a very concrete example use case, then discuss whether that use case is common enough to add it to the client library.

**Comment by adamzhoul** [comment]
> I think it's helpful to have a very concrete example use case, then discuss whether that use case is common enough to add it to the client library.

what about: `use index`? itself is a concrete use case. 

Using index is very common, for example:
1. using nodeName index in the monitoring app. to get pods in a specific node. ( nodeName=xxx)
2. using label index in the scheduling app.  to calculate how many pods already exist of a specific kind. (apptype=xxxx)
3. using label index in PaaS platform. to show developers all pods they have deployed. (app=xxx)
4. ...

Each use case can use `podInformer.Informer().GetIndexer().ByIndex(indexName, key)` 
But, using both `inform` and `xxxLister` to get objects is weird. using `xxxLister` only is better.
However,  `xxxLister` wrapped the index, it can't be reached outside.


Besides:
`xxxLister` has extension `XxxListerExpansion `. I think the designer wants to expand its usage.
 But don't know how to use it.  do you know?  @yliaog


thanks.

**Comment by yliaog** [comment]
> Each use case can use `podInformer.Informer().GetIndexer().ByIndex(indexName, key)`
> But, using both `inform` and `xxxLister` to get objects is weird. using `xxxLister` only is better.
> However, `xxxLister` wrapped the index, it can't be reached outside.

client library utils, e.g Lister, Indexer, Informer are often used together to solve a certain problem. could you give an example that uses Lister, or Indexer alone? It would be helpful to have an "end to end" example code to look at.

**Comment by adamzhoul** [comment]
> It would be helpful to have an "end to end" example code to look at.

## Platform architecture:             
Paas Platform  ------ HTTP call -------->  middleWare application ------------> k8s apiServer

## Background:
1. on PaasPlatform, we have a page to show all instances of an app. 
2. we have serval deployments, like a different version, different env. so we use a label selector to get all pods.
3. PaasPlatform is not calling k8s directly, all calls to k8s is in the middleware application. 

## The core is how we write the middleware application.

```golang

// main.go
func main(){

   // 1. start a controller with informer inside.  
   podController := pod.NewController(indexs)
   podController.Run() // run informer to sync pods , actually.
   defer podController.Stop()

   // 2. start an HTTP Server, handle crud.
   // pass in a Lister is enough and it has a more reasonable interface.
   // pass informer is too powerful for handlers. 
   // and informer is the concept of k8s, the HTTP package should know the less the better.
   web.HTTPStart(podController.Lister)  

   ...
}

// pod/controller.go
type Controller struct{
    Lister podLister
    ...
}

func NewController(indexers cache.Indexers) (*Controller, error) {

   // New Informer
   // add indexers to infromer
   // generator a Lister by informer.index 
   
}


// web/httpHandler.go
func ListPods(w http.ResponseWriter, req *http.Request) {

    podLister.ListByIndex(....)  // 
}

```


In all:
  Informer is too powerful for other modules to use. 
  Lister on the contract is clearly telling the caller, what it is and what you should do with it. 
  And Has limited interface and responsibility. 
  After all, we are using the client library not writing k8s code.

**Comment by yliaog** [comment]
In your example, you're passing a pod lister to the handler. Instead, you could pass an indexer to the handler, and inside the handler, you can select which index to use with something like indexer.ByIndex("nodeName", node.Name). then you may run label selector through the results to pick the matches.

PodNamespaceLister is a special lister with Namespace Index, as mentioned previously, namespace is special in kubernetes, hence the library adds such a special lister.

If you'd like to generalize that to pod lister with arbitrary index, and add it to the client library, i think the best way forward is to ask the community on how many people would need it in the client library. could you send a survey email to the community mailing list kubernetes-dev@?

**Comment by adamzhoul** [comment]
thanks, @yliaog. 

Anyway, there is not an answer, which one is better.
It's up to the situation or developer's habit or a different perspective on those components, their responsibility, or else. 


Follow your advice, I will send an email later. thanks.

**Comment by adamzhoul** [comment]
Not working on this recently
/close

**Pull request closed without merging**